📖 User guide: Add nodeReuse feature document (#246)
metal3-io/metal3-docs: docs/user-guide/src/capm3/node_reuse.md

# Node Reuse

This feature brings a possibility of re-using the same BaremetalHosts (referred to as a host later)

> during deprovisioning/provisioning mainly as part of the rolling upgrade process in the cluster.

Reviewer: Is this part trying to answer _why we need it?_  just thinking that we might need to emphasize that need for using the same external storage

Author: What is the external storage in this context, do you mean m3 object, i.e m3mt?

Reviewer: I mean as a necessity that node reuse cover. as we have an external storage attached to a node we prefer reusing the same node and keep the external storage I guess that is one of the main reason to implement nodereuse, isn't it?

Author: Yeah, but that is a motivation of the feature which is described in detail in the original proposal https://github.com/metal3-io/metal3-docs/blob/main/design/cluster-api-provider-metal3/node_reuse.md#motivation.

## Workflow

Metal3MachineTemplate (M3MT) CRD is the object to support enabling/disabling node reuse feature. When set to **true**,
Cluster API Provider Metal3 (CAPM3) Machine controller performs the following:

During deprovisioning:

1. Finds M3MT referenced by Metal3Machine (M3M) and checks if nodeReuse is enabled;
1. If enabled, it checks Cluster API (CAPI) machine object role (control-plane or worker), and based on that, fetches
   either KubeadmControlPlane (KCP) or MachineDeployment (MD) name accordingly;

> 1. Sets fetched object's name on the corresponding host's labels field.

Reviewer: Would be better to explain the label format here: <infrastructure.cluster.x-k8s.io/node-reuse: kcp-object-name>

Author: I would keep it as is since it is already detailed and more than that can be found in the original proposal of the feature.

> On the following provisioning:

Reviewer 2: ```suggestion
During the next provisioning:
```

1. Filters out all hosts with node reuse label set and those that store KCP/MD object name, in a format of i.e:
  **<infrastructure.cluster.x-k8s.io/node-reuse: kcp-object-name>** where after node reuse label, a prefix is
  appended depending on the owner of the host in question (kcp or md) to avoid mismatch, in cases where KCP and
  MD object names are the same.
1. Chooses a host from the filtered list of reused hosts for provisioning.

> 

Reviewer: Can we project the steps on an example ?

Author: I am not sure if I got the idea, can you elaborate more?

Reviewer: I was thinking if an example that applies those steps on a list of bmh would simplify understanding the node reuse workflow

Author: @mboukhalfa having a controller logic re-explained on this document felt over-burden, so I have completely removed the controller logic from the user guide and provided a reference to it, thus I have removed the points we are discussing here

## Importance of scale-in strategy

The logic behind the reusing of the hosts, solely relies on the scale-in upgrade strategy utilized by
CAPI objects, i.e KCP or MachineDeployment, where during the upgrade process, the machines owned

> (control-plane or worker) are removed one-by-one before creating a new ones (delete-create method).

Reviewer 2: ```suggestion
(control-plane or worker) are removed one-by-one before creating new ones (delete-create method).
```

That way, we can fully ensure that, once the upgrade is kicked in, the intended host is reused
during deprovisioning and the same host is picked up on the following provisioning for the new machine
being created.

**Note:** To achieve the desired *delete first and create after* like behavior in above-mentioned CAPI objects,
user has to modify:

* MaxSurge field and set it to 0 during KCP upgrade with minimum of 3 replicas of control plane machines in the

>   cluster, so that control plane machines are removed one-by-one before creating a new ones;

Reviewer 2: ```suggestion
  cluster, so that control plane machines are removed one-by-one before creating new ones;
```

# Node Reuse

This feature brings a possibility of re-using the same BaremetalHosts (referred to as a host later)

> during deprovisioning and provisioning mainly as part of the rolling upgrade process in the cluster.

Reviewer 3: ```suggestion
during deprovisioning and provisioning as a part of the rolling upgrade process in the cluster.
```

## Importance of scale-in strategy

The logic behind the reusing of the hosts, solely relies on the **scale-in** upgrade strategy utilized by

> Cluster API objects, namely [KubeadmControlPlane](https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20191017-kubeadm-based-control-plane.md#rolling-update-strategy) and MachineDeployment, where during the upgrade process, the machines owned (control-plane or worker) are removed one-by-one before creating new ones (delete-create method).

Reviewer 3: Can you please break this long sentence? It would be easier for user to understand.

> That way, we can fully ensure that, once the upgrade is kicked in, the intended host is reused (picked up on the following provisioning for the new machine being created).

Reviewer 3: ```suggestion
In this way, we can ensure the intended host is reused when upgrade is kicked in (picked up on the following provisioning for the new machine being created).
```

> Cluster API objects, namely [KubeadmControlPlane](https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20191017-kubeadm-based-control-plane.md#rolling-update-strategy) and MachineDeployment, where during the upgrade process, the machines owned (control-plane or worker) are removed one-by-one before creating new ones (delete-create method).
> That way, we can fully ensure that, once the upgrade is kicked in, the intended host is reused during deprovisioning and the
> same host is picked up on the following provisioning for the new machine being created.

Reviewer 4: ```suggestion
Cluster API objects, namely [KubeadmControlPlane](https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20191017-kubeadm-based-control-plane.md#rolling-update-strategy) and MachineDeployment, where during the upgrade process, the machines owned (control-plane or worker) are removed one-by-one before creating new ones (delete-create method).
That way, we can fully ensure that, once the upgrade is kicked in, the intended host is reused (picked up on the following provisioning for the new machine being created).
```

> **Note:** To achieve the desired *delete first and create after* like behavior in above-mentioned Cluster API objects,

Reviewer 3: ```suggestion
**Note:** To achieve the desired *delete first and create after*  behavior in above-mentioned Cluster API objects,
```

**Note:** To achieve the desired *delete first and create after* like behavior in above-mentioned Cluster API objects,
user has to modify:

* MaxSurge field in KubeadmControlPlane and set it to 0 with minimum number of 3 control plane machines replicas
* MaxSurge and MaxUnavailable fields in MachineDeployment set them to 0 & 1 accordingly

On the contrary, if the scale-out strategy is utilized by CAPI objects during the upgrade, usually create-swap-delete
method is followed by CAPI objects, where new machine is created first and new host is picked up for that
machine, breaking the node reuse logic right at the beginning of the upgrade process.

## Workflow

Metal3MachineTemplate (M3MT) Custom Resource is the object responsible for enabling of the node reuse feature.

```yaml
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: Metal3MachineTemplate
metadata:
  name: test1-controlplane
  namespace: metal3
spec:
  nodeReuse: True
  template:
    spec:
      image:
      ...
```

> There are two Metal3MachineTemplate objects, one referenced by KubeadmControlPlane which is used to create Metal3Machine owned by control plane machine, and the other one referenced by MachineDeployment and used to create Metal3Machine owned by worker machine. Before performing an upgrade, user must set `nodeReuse` field to **true** in the desired Metal3MachineTemplate object where hosts targeted to be reused. If left unchanged, by default, `nodeReuse` field is set to **false** meaning no host reusing performed in the workflow.
> 

Reviewer 4: ```suggestion
There could be two Metal3MachineTemplate objects, one referenced by KubeadmControlPlane for control plane nodes, and the other by MachineDeployment for worker node. Before performing an upgrade, user must set `nodeReuse` field to **true** in the desired Metal3MachineTemplate object where hosts targeted to be reused. If left unchanged, by default, `nodeReuse` field is set to **false** meaning no host reusing performed in the workflow.

```

> Once `nodeReuse` field is set to **true**, user has to make sure that scale-in feature is enabled as suggested above, and go on with updating the desired fields in KubeadmControlPlane or MachineDeployment to start a rolling upgrade.

Reviewer 4: ```suggestion
Once `nodeReuse` field is set to **true**, user has to make sure that scale-in feature is enabled as suggested above, and goes on with updating the desired fields in KubeadmControlPlane or MachineDeployment to start a rolling upgrade.
```
